### PR TITLE
Cluster-autoscaler: decrease target size function in cloud provider interface

### PR DESCRIPTION
**Pull request by mwielgus**
Sometimes the cloud provider has troubles with providing the requested amount of nodes either due to quota or to (temporary) technical limitations. In such a case CA should be able to change the previously made request without killing any of the existing nodes. 

@andrewsykim - will this PR work on AWS?

cc: @jszczepkowski @andrewsykim @fgrzadkowski

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go [hunk 1]
```diff
@@ -142,6 +142,30 @@ func (asg *Asg) IncreaseSize(delta int) error {
...
+	size, err := asg.awsManager.GetAsgSize(asg)
+	if err != nil {
+		return err
+	}
+	nodes, err := asg.awsManager.GetAsgNodes(asg)
+	if err != nil {
+		return err
+	}
+	if int(size)+delta < len(nodes) {
+		return fmt.Errorf("attempt to delete existing nodes targetSize:%d delta:%d existintNodes: %d",
```
Comment (andrewsykim) [inline cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go]: `s/existintNodes/existingNodes`

File: cluster-autoscaler/cloudprovider/gce/gce_cloud_provider.go [hunk 2]
```diff
@@ -146,6 +146,28 @@ func (mig *Mig) IncreaseSize(delta int) error {
 	return mig.gceManager.SetMigSize(mig, size+int64(delta))
 }
 
+// DecreaseTargetSize decreases the target size of the node group. This function
+// doesn't permit to delete any existing node and can be used only to reduce the
+// request for new nodes that have not been yet fulfilled. Delta should be negative.
+func (mig *Mig) DecreaseTargetSize(delta int) error {
+	if delta <= 0 {
```
Comment (andrewsykim) [inline cluster-autoscaler/cloudprovider/gce/gce_cloud_provider.go]: should be `delta >= 0 `?

**Review by andrewsykim — COMMENTED** [review]
@mwielgus I don't think this approach would work on AWS as well as it would on GCE. Though from my experience its rare to get rate limited for ec2 instances on AWS unless you're on spot instances. 

If you increase the ASG size from 5 -> 10 and end up with 8 nodes after the timeout and then set the desired size back to 8 there could be a chance the autoscaling group may attempt to delete nodes from the existing 8 nodes. There are a lot of edge cases to consider. I'll get into contact with some AWS support people to confirm.

File: cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go (hunk 1 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go]: Done.

File: cluster-autoscaler/cloudprovider/gce/gce_cloud_provider.go (hunk 2 above)
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/gce/gce_cloud_provider.go]: Yes, obviously. Thanks for catching this.

**Comment by mwielgus** [comment]
@andrewsykim - I made some experiments/checks on GCE and it seems to delete "uncreated" instances first. I will double check with people who actually wrote this code but it seems to be a reasonable strategy to recover from quota overrun. How otherwise it should be done?

**Comment by andrewsykim** [comment]
@mwielgus I discussed this with AWS support. Decreasing the ASG size would work for the most part if there are no instances launching. I'm still a bit hesitate because its not entirely clear when a instance is "launching" in the AWS world (when the request was received vs when the VM is actually launching). There's a race condition to consider, it seems possible that you could decrease the ASG size as soon as the instances you waited for launched, in which case its very likely that you would terminate an existing node (with no drain). However, this race condition seems possible in GCE too right? There are a few alternatives I can think of at the moment but it would involve a fair bit of refactoring. Given that its unlikely, it might be worth it to go forward with the current implementation and reconsider those alternatives once we understand the likelyhood of those race conditions.

**Comment by mwielgus** [comment]
Yes, I'm aware of the potential race condition here. To mitigate the risk we will only do it if the difference between the number of started nodes and the target was constant for a prolonged time (like 15 min). 

What alternatives do you have on your mind?

**Comment by mwielgus** [comment]
On GCE you can find all instances managed by a node-group and delete the uncreated one 

$ gcloud compute instance-groups managed list-instances gke-cluster-1-default-pool-96a2740c-grp 

NAME                                      ZONE            STATUS   ACTION    LAST_ERROR
gke-cluster-1-default-pool-96a2740c-13xf  europe-west1-c           CREATING  Error QUOTA_EXCEEDED: Instance 'gke-cluster-1-default-pool-96a2740c-13xf' creation failed: Quota 'IN_USE_ADDRESSES' exceeded.  Limit: 23.0

$ gcloud compute instance-groups managed delete-instances gke-cluster-1-default-pool-96a2740c-grp --instances=gke-cluster-1-default-pool-96a2740c-48g8

Is there such a way in AWS?

**Comment by andrewsykim** [comment]
@mwielgus not that I'm aware of. The alternative I had in mind was enabling [instance protection](http://docs.aws.amazon.com/autoscaling/latest/userguide/as-instance-termination.html#instance-protection). Only instances that are `InService` can qualify for instance protection and we can enable/disable instance protection for individual nodes. If we enable instance protection for any node visible to CA, it can't be terminated by decreasing the desired size unless we explicity do the [TerminateInstance](https://github.com/kubernetes/contrib/blob/master/cluster-autoscaler/cloudprovider/aws/aws_manager.go#L155) operation.

**Comment by mwielgus** [comment]
@andrewsykim What about turning on instance protection automatically on all registered/ready nodes?

**Comment by mwielgus** [comment]
@andrewsykim Well that what we want there - to delete the nodes about which we are sure that we want to delete them. Do you think that such a side effect (automatic protection on all registered/ready) nodes would be acceptable for the AWS users?

**Comment by andrewsykim** [comment]
At first thought, if its well documented and there's logging for it on the CA, then I don't see it being a problem. But let me think about it and ask others for their opinions. :)

**Comment by jszczepkowski** [comment]
LGTM

**Pull request merged**